File: ticktock-compiler/src/test/kotlin/dev/zacsweers/ticktock/compiler/JavaWriterTest.java
package dev.zacsweers.ticktock.compiler;

import static com.google.testing.compile.CompilationSubject.assertThat;
import static com.google.testing.compile.Compiler.javac;
import static com.google.testing.compile.JavaFileObjectSubject.assertThat;

import com.google.common.base.Charsets;
import com.google.testing.compile.Compilation;
import com.google.testing.compile.JavaFileObjects;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.LinkedHashSet;
import javax.tools.JavaFileObject;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class) public class JavaWriterTest {

  @Rule public TemporaryFolder tmpFolder = new TemporaryFolder();

  private static final String SOURCE_NAME = "ticktock/LazyZoneRules";

  private Path outputDir;
  private JavaWriter javaWriter;

  @Before public void setup() throws IOException {
    outputDir = tmpFolder.newFolder()
        .toPath();
    javaWriter = new JavaWriter(outputDir);
  }

  private JavaFileObject generatedSource(String version, String... zoneIds) throws Exception {
    javaWriter.writeZoneIds("ticktock",
        version,
        new LinkedHashSet<>(Arrays.asList(zoneIds)));
    Path output = outputDir.resolve(SOURCE_NAME + ".java");
    String sourceString = new String(Files.readAllBytes(output), Charsets.UTF_8);
    return JavaFileObjects.forSourceString(SOURCE_NAME, sourceString);
  }

  @Test public void writeZoneIds() throws Exception {
    JavaFileObject source = generatedSource("2010a", "Europe/Berlin", "UTC", "US/Pacific");
    JavaFileObject expected = JavaFileObjects.forSourceString(SOURCE_NAME,
        ""
            + "package ticktock;"
            + "\n"
            + "import java.lang.String;"
            + "import java.util.Arrays;"
            + "import java.util.List;"
            + "\n"
            + "final class LazyZoneRules {\n"
            + "    static final String VERSION = \"2010a\";\n"
            + "\n"
            + "    static final List<String> REGION_IDS = Arrays.asList(\n"
            + "            \"Europe/Berlin\",\n"
            + "            \"UTC\",\n"
            + "            \"US/Pacific\");\n"
            + "        }");

    Compilation compilation = javac().compile(source);
    assertThat(compilation).succeeded();
    assertThat(source).hasSourceEquivalentTo(expected);
  }
}

File: ticktock-compiler/src/test/kotlin/dev/zacsweers/ticktock/compiler/JavaWriterTest.kt
/*
 * Copyright (C) 2020 Zac Sweers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.zacsweers.ticktock.compiler

import com.google.testing.compile.CompilationSubject
import com.google.testing.compile.Compiler
import com.google.testing.compile.JavaFileObjectSubject
import com.google.testing.compile.JavaFileObjects
import java.nio.file.Files
import java.nio.file.Path
import javax.tools.JavaFileObject
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class JavaWriterTest {

  @get:Rule
  val tmpFolder = TemporaryFolder()

  private lateinit var outputDir: Path
  private lateinit var javaWriter: JavaWriter

  @Before
  fun setup() {
    outputDir = tmpFolder.newFolder().toPath()
    javaWriter = JavaWriter(outputDir)
  }

  private fun generatedSource(version: String, vararg zoneIds: String): JavaFileObject {
    javaWriter.writeZoneIds("ticktock", version, setOf(*zoneIds))
    val output = outputDir.resolve("$SOURCE_NAME.java")
    val sourceString = Files.readAllBytes(output).decodeToString()
    return JavaFileObjects.forSourceString(SOURCE_NAME, sourceString)
  }

  @Test
  fun writeZoneIds() {
    val source = generatedSource("2010a", "Europe/Berlin", "UTC", "US/Pacific")
    val expected = JavaFileObjects.forSourceString(
        SOURCE_NAME,
        """package ticktock;

import dev.zacsweers.ticktock.runtime.ZoneIdsProvider;
import java.lang.Override;
import java.lang.String;
import java.util.Arrays;
import java.util.List;

final class GeneratedZoneIdsProvider implements ZoneIdsProvider {
  private static final String VERSION_ID = "2010a";

  private static final List<String> ZONE_IDS = Arrays.asList(
      "Europe/Berlin",
      "UTC",
      "US/Pacific");

  @Override
  public String getVersionId() {
    return VERSION_ID;
  }

  @Override
  public List<String> getZoneIds() {
    return ZONE_IDS;
  }
}""")
    val compilation = Compiler.javac().compile(source)
    CompilationSubject.assertThat(compilation).succeeded()
    JavaFileObjectSubject.assertThat(source).hasSourceEquivalentTo(expected)
  }

  companion object {
    private const val SOURCE_NAME = "ticktock/GeneratedZoneIdsProvider"
  }
}
